demos/guide/quic-hq-interop.c

```diff
@@ -0,0 +1,1109 @@
+        (void)BIO_flush(bio_keylog);
+    }
+    SSL_CTX_set_keylog_callback(ctx, keylog_callback);
+    return 0;
+}
+
+/**
+ * @brief A static integer indicating whether the session is cached.
+ *
+ * This variable is used to track the state of session caching. It is
+ * initialized to 0, meaning no session is cached. The value may be updated
+ * to indicate that a session has been successfully cached.
+ *
+ * @note This variable is static, meaning it is only accessible within the
+ * file in which it is declared.
+ */
+static int session_cached = 0;
+
+/**
+ * @brief Caches a new SSL session if one is not already cached.
+ *
+ * This function writes a new SSL session to the session BIO and caches it.
+ * It ensures that only one session is cached at a time by checking the
+ * `session_cached` flag. If a session is already cached, the function
+ * returns without caching the new session.
+ *
+ * @param ssl A pointer to the SSL object associated with the session.
+ * @param sess A pointer to the SSL_SESSION object to be cached.
+ * @return 1 if the session is successfully cached, 0 otherwise.
+ *
+ * @note This function only allows one session to be cached. Subsequent
+ * sessions will not be cached unless `session_cached` is reset.
+ */
+static int cache_new_session(struct ssl_st *ssl, SSL_SESSION *sess)
+{
+
+    if (session_cached == 1)
+        return 0;
+
+    /* Just write the new session to our bio */
+    if (!PEM_write_bio_SSL_SESSION(session_bio, sess))
+        return 0;
+
+    (void)BIO_flush(session_bio);
+    /* only cache one session */
+    session_cached = 1;
+    return 1;
+}
+
+/**
+ * @brief Sets up the session cache for the SSL connection.
+ *
+ * This function configures session caching for the given SSL connection
+ * and context. It attempts to load a session from the specified cache file
+ * or creates a new one if the file does not exist. It also configures the
+ * session cache mode and disables stateless session tickets.
+ *
+ * @param ssl A pointer to the SSL object for the connection.
+ * @param ctx A pointer to the SSL_CTX object representing the context.
+ * @param filename The name of the file used to store the session cache.
+ * @return 1 on success, 0 on failure.
+ *
+ * @note If the cache file does not exist, a new file is created and the
+ * session cache is initialized. If a session is successfully loaded from
+ * the file, it is added to the context and set for the SSL connection.
+ * If an error occurs during setup, the session BIO is freed.
+ */
+static int setup_session_cache(SSL *ssl, SSL_CTX *ctx, const char *filename)
+{
+    SSL_SESSION *sess = NULL;
+    int rc = 0;
+    int new_cache = 0;
+
+    /* make sure caching is enabled */
+    if (!SSL_CTX_set_session_cache_mode(ctx, SSL_SESS_CACHE_BOTH))
+        return rc;
+
+    /* Don't use stateless session tickets */
+    if (!SSL_CTX_set_options(ctx, SSL_OP_NO_TICKET))
+        return rc;
```

Comment:
I am curious why (stateless) session tickets are disabled. Are they deprecated in QUIC? Or best avoided in the interop test? ... In any case, this call has no effect on the client side...

Reply:
IIRC this was done because several servers tested with the interop harness encountered errors when using stateless session tickets, and so I did this to disable them.  I admit to not looking to closely at the errors when the problem was resolved.  I can try removing it.